neat_core.c

```diff
@@ -804,6 +1035,210 @@ neat_open(neat_ctx *mgr, neat_flow *flow, const char *name, uint16_t port)
+                    "Call to setsockopt failed with errno=%d", errno);
+            return NEAT_ERROR_IO;
+        }
+#endif
+        return NEAT_ERROR_OK;
+    }
+#endif // !(defined(__FreeBSD__) || defined(__NetBSD__) || defined (__APPLE__))
+
+
+    return NEAT_ERROR_UNABLE;
+}
+
+static void
+set_primary_dest_resolve_cb(struct neat_resolver *resolver, struct neat_resolver_results *results, uint8_t code)
+{
+    neat_flow *flow = (neat_flow *)resolver->userData1;
+    struct neat_ctx *ctx = flow->ctx;
+    neat_log(NEAT_LOG_DEBUG, "%s", __func__);
+
+    if (code != NEAT_RESOLVER_OK) {
+        io_error(ctx, flow, code);
+        return;
+    }
+
+    assert(results->lh_first);
+
+#ifdef USRSCTP_SUPPORT
+    struct sctp_setprim addr;
+    addr.ssp_addr = results->lh_first->dst_addr;
+
+    if (usrsctp_setsockopt(flow->sock, IPPROTO_SCTP, SCTP_PRIMARY_ADDR, &addr, sizeof(addr)) < 0) {
+        neat_log(NEAT_LOG_DEBUG, "Call to usrsctp_setsockopt failed");
+        return;
+    }
+#elif defined(HAVE_NETINET_SCTP_H) && defined(__linux__)
+    struct sctp_prim addr;
+    struct sctp_assocparams assocparams;
+    unsigned int optlen = sizeof(assocparams);
+    // addr.ssp_addr = results->lh_first->dst_addr;
+    int rc = getsockopt(flow->fd, IPPROTO_SCTP, SCTP_ASSOCINFO, &assocparams, &optlen);
+    if (rc < 0) {
+        neat_log(NEAT_LOG_DEBUG, "Call to getsockopt failed");
+        return;
+    }
+
+    neat_log(NEAT_LOG_DEBUG, "assoc: %d", assocparams.sasoc_assoc_id);
+
+    addr.ssp_assoc_id = assocparams.sasoc_assoc_id;
+    addr.ssp_addr = results->lh_first->dst_addr;
+
+    rc = setsockopt(flow->fd, IPPROTO_SCTP, SCTP_PRIMARY_ADDR, &addr, sizeof(addr));
+    if (rc < 0) {
+        neat_log(NEAT_LOG_DEBUG, "Call to setsockopt failed");
+        return;
+    }
+#endif
+
+    char dest_addr[INET6_ADDRSTRLEN];
+    inet_ntop(results->lh_first->ai_family,
+            (const void*)&(results->lh_first->dst_addr),
+            dest_addr,
+            INET6_ADDRSTRLEN);
+
+    struct sockaddr_storage* sockaddr = &results->lh_first->dst_addr;
+
+    if (results->lh_first->ai_family == AF_INET) {
+        inet_ntop(AF_INET, &(((struct sockaddr_in *) (sockaddr))->sin_addr), dest_addr, INET6_ADDRSTRLEN);
+    } else {
+        inet_ntop(AF_INET6, &(((struct sockaddr_in6 *) (sockaddr))->sin6_addr), dest_addr, INET6_ADDRSTRLEN);
+    }
+
+    neat_log(NEAT_LOG_DEBUG, "Updated primary destination address to: %s", dest_addr);
+}
+
+neat_error_code
+neat_set_primary_dest(struct neat_ctx *ctx, struct neat_flow *flow, const char *name)
+{
+    neat_log(NEAT_LOG_DEBUG, "%s", __func__);
+
+#if defined(IPPROTO_SCTP)
+    if (flow->sockProtocol == IPPROTO_SCTP) {
+        int protocols[] = {IPPROTO_SCTP};
+
+        ctx->resolver->handle_resolve = set_primary_dest_resolve_cb;
+        neat_getaddrinfo(ctx->resolver, AF_UNSPEC, name, flow->port,
+                protocols, sizeof(*protocols)/sizeof(protocols[0]));
+
```

Comment:
Not sure if it would make sense to provide there more than one protocol candidate. But does `neat_getaddrinfo` ensure that name is an IP-address? I think the argument `name` must be one...